tsdb/chunkenc/histo.go

```diff
@@ -0,0 +1,711 @@
+	// When using striped locks to guard access to chunks, probably yes.
+	// Could only copy data if the chunk is not completed yet.
+	//if histoIter, ok := it.(*histoIterator); ok {
+	//	histoIter.Reset(c.b.bytes())
+	//	return histoIter
+	//}
+	return &histoIterator{
+		// The first 2 bytes contain chunk headers.
+		// We skip that for actual samples.
+		br:       newBReader(c.b.bytes()[2:]),
+		numTotal: binary.BigEndian.Uint16(c.b.bytes()),
+		t:        math.MinInt64,
+	}
+}
+
+// Iterator implements the Chunk interface.
+//func (c *HistoChunk) Iterator(it Iterator) Iterator {
+//	return c.iterator(it)
+//}
+
+type histoAppender struct {
+	b *bstream
+
+	// Meta
+	schema             int32
+	posSpans, negSpans []histogram.Span
+
+	// for the fields that are tracked as dod's
+	t                           int64
+	cnt, zcnt                   uint64
+	tDelta, cntDelta, zcntDelta uint64
+
+	posbuckets, negbuckets           []int64
+	posbucketsDelta, negbucketsDelta []int64 // these could probably be uints instead. if the counts go up, the delta will always be >=0
+
+	// for the fields that are gorilla xor coded
+	sum      float64
+	leading  uint8
+	trailing uint8
+
+	buf64 []byte // for working on varint64's
+}
+
+func putVarint(b *bstream, buf []byte, x int64) {
+	for _, byt := range buf[:binary.PutVarint(buf, x)] {
+		b.writeByte(byt)
+	}
+}
+
+func putUvarint(b *bstream, buf []byte, x uint64) {
+	for _, byt := range buf[:binary.PutUvarint(buf, x)] {
+		b.writeByte(byt)
+	}
+}
+
+// we use this for millisec timestamps and all counts
+// for now this is copied from xor.go - we will probably want to be more conservative (use fewer bits for small values) - can be tweaked later
+func putDod(b *bstream, dod int64) {
+	switch {
+	case dod == 0:
+		b.writeBit(zero)
+	case bitRange(dod, 14):
+		b.writeBits(0x02, 2) // '10'
+		b.writeBits(uint64(dod), 14)
+	case bitRange(dod, 17):
+		b.writeBits(0x06, 3) // '110'
+		b.writeBits(uint64(dod), 17)
+	case bitRange(dod, 20):
+		b.writeBits(0x0e, 4) // '1110'
+		b.writeBits(uint64(dod), 20)
+	default:
+		b.writeBits(0x0f, 4) // '1111'
+		b.writeBits(uint64(dod), 64)
+	}
+}
+
+func (a *histoAppender) Append(int64, float64) {
+	panic("cannot call histoAppender.Append().")
+}
+
+func (a *histoAppender) AppendHistogram(h histogram.SparseHistogram) {
+	var tDelta, cntDelta, zcntDelta uint64
+	num := binary.BigEndian.Uint16(a.b.bytes())
+
+	if num == 0 {
+		putVarint(a.b, a.buf64, h.Ts)
+		putUvarint(a.b, a.buf64, h.Count)
+		putUvarint(a.b, a.buf64, h.ZeroCount)
+		a.b.writeBits(math.Float64bits(h.Sum), 64)
+		for _, buck := range h.PositiveBuckets {
+			putVarint(a.b, a.buf64, buck)
+		}
+		for _, buck := range h.NegativeBuckets {
+			putVarint(a.b, a.buf64, buck)
+		}
+	} else if num == 1 {
+		tDelta = uint64(h.Ts - a.t)
+
+		// WARNING: we assume all counts go up. what guarantee do we have this is true? uint may underflow if not.
```

Comment:
I think we should be able to handle counts going down (AKA counter reset), even if we (for now) decide to start a new chunk on resets.

Reply:
ok, so we make the delta types all int64. can we then also change `SparseHistogram.{Count,ZeroCount}` to int64 ? 
so timestamp, count and zerocount, as well as all their delta's would be int64 types, removing the need to cast back and forth

@codesome what do you think?

Reply:
Sounds fine to me, we can add comments to variables when it can be negative and when it has to be always positive.